l2geth/core/state_transition_ovm.go

```diff
@@ -44,12 +45,13 @@ func toExecutionManagerRun(evm *vm.EVM, msg Message) (Message, error) {
 		return nil, err
 	}
 
+	gasLimit := fees.DecodeL2GasLimitU64(msg.Gas())
 	outputmsg, err := modMessage(
 		msg,
 		msg.From(),
 		&evm.Context.OvmExecutionManager.Address,
 		ret,
-		evm.Context.GasLimit,
+		gasLimit,
```

Comment:
I believe this is actually incorrect. We want to pass *all* available gas to the transaction but then only pass the L2 gas within the `ECDSAContractAccount` (which you've done below). Otherwise the `gasleft() > gasLimit` check you're doing in the contract account will always fail.

Reply:
Doing this would need to be accompanied by two additional changes in the ECDSA contract account if I understand correctly:


To prevent a malicious contract from using more gas than what was changed for:
```
                return transaction.to.call{gas: gasLimit}(transaction.data);
```
https://github.com/ethereum-optimism/optimism/blob/b355be0b23ab80d9cdcb0b1f481eaeec02412b20/packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol#L143

Same thing for OVMCREATE:
```
 (address created, bytes memory revertdata) = Lib_ExecutionManagerWrapper.ovmCREATE( 
     transaction.data,
     gasLimit 
 ); 
```

https://github.com/ethereum-optimism/optimism/blob/b355be0b23ab80d9cdcb0b1f481eaeec02412b20/packages/contracts/contracts/optimistic-ethereum/OVM/accounts/OVM_ECDSAContractAccount.sol#L110-L112

Reply:
Before the contract account passes through a specific gasLimit to its ovmCALL or ovmCREATE, it should subtract the intrinsic gas cost which is defined by https://github.com/ethereum/go-ethereum/blob/2d716c4b01037f7f7c96795784909ec8230ac523/core/state_transition.go#L115

Can just take the calldata size and multiply by 16 to simplify the need to iterate and count in the evm